lang/cs/Org.Apache.REEF.IMRU/OnREEF/StateMachine/SystemStateMachine.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+// 
+//   http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using Org.Apache.REEF.IMRU.OnREEF.Exceptions;
+
+namespace Org.Apache.REEF.IMRU.OnREEF.StateMachine
+{
+    /// <summary>
+    /// This is a state transition machine which wraps current state, defines valid state transitions and API to move to next state. 
+    /// All the system states are define in <see cref="SystemState"></see>
+    /// For the system state transition diagram <see href="https://issues.apache.org/jira/browse/REEF-1327"></see>
+    /// </summary>
+    internal sealed class SystemStateMachine
+    {
+        internal readonly static IDictionary<StateTransition<SystemState, SystemStateEvent>, SystemState> Transitions = 
+            new ReadOnlyDictionary<StateTransition<SystemState, SystemStateEvent>, SystemState>(
+                new Dictionary<StateTransition<SystemState, SystemStateEvent>, SystemState>()
+        {
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.WaitingForEvaluator, SystemStateEvent.AllContextsAreReady), SystemState.SubmittingTasks },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.WaitingForEvaluator, SystemStateEvent.NotRecoverable), SystemState.Fail },
```

Comment:
There are many evaluator failures that appen between IAllocatedEvlauator and IActiveContext. The system will be in  ystemState.WaitingForEvaluator state and event will be FailedEvaluator. This case is not covered here and probably in design also. We should asks for failed evaluator again in this case if root context did not fail because of user error in downloading data or other actions in Input partition. Note that data download will happen in root context, so any failure there will lead to Failed evaluator. So we need to distinguish between actual system falure and user error in root context.